### PR TITLE
Improvements to `bundle add`, plus `bundle remove`

### PR DESCRIPTION
**Pull request by shime**
Adding the RFC for `bundle add` improvements and `bundle remove` command, as described in [shime/bun#3](https://github.com/shime/bun/issues/3).

Looking forward to your input!

[Rendered RFC text](https://github.com/shime/rfcs/blob/bundle-add-improvements/text/0002-bundle-add-and-remove.md)

### CODE REVIEW

File: text/0002-bundle-add-and-remove.md [hunk 1]
```diff
@@ -0,0 +1,275 @@
...
+```ruby
+bundle add minitest --group test
+```
+
+```ruby
+# Gemfile after
+source "https://rubygems.org"
+
+group :test do
+  gem "minitest", "~> 5.11"
```
Comment (colby-swandale) [inline text/0002-bundle-add-and-remove.md]: I don't agree with adding a version constraint when it wasn't specified by the user. The version constraint should be defined  by the `--version` option that `bundle add` has already.

Comment (shime) [reply text/0002-bundle-add-and-remove.md]: `bundle add` already works this way if you don't specify the version. How should it behave if the version is not specified?

Comment (colby-swandale) [reply text/0002-bundle-add-and-remove.md]: ah ok, didn't know that. nevermind then

File: text/0002-bundle-add-and-remove.md [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+```
+
+```ruby
+# Gemfile after
+source "https://rubygems.org"
+```
+
+### Description
+
+The `remove` command removes gems from the Gemfile, along with their corresponding group block if it remains empty after the deletion. By default it runs `bundle install` afterwards, which can also be skipped with the `--skip-install` flag.
```
Comment (segiddins) [inline text/0002-bundle-add-and-remove.md]: It is `remove` or `rm`? I'm in favor of `remove`, for symmetry with `add`

File: text/0002-bundle-add-and-remove.md (hunk 1 above)
Comment (hmistry) [reply text/0002-bundle-add-and-remove.md]: @colby-swandale FYI, this was @indirect default requirement in the original `add` command. He wanted conservative versioning to be set as it's the safest and can be overridden if the user added a specific version.

File: text/0002-bundle-add-and-remove.md (hunk 2 above)
Comment (hmistry) [reply text/0002-bundle-add-and-remove.md]: I too am in favour of `remove` for same symmetry reason and also most bundler commands are whole words, not abbreviations/shortened versions. User will therefore expect whole words as commands.

File: text/0002-bundle-add-and-remove.md [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+```
+
+```
+bundle add minitest sinatra
+```
+
+Prints out and exits with error status code:
+
+```shell
+Gem "minitest" already present in the Gemfile. Aborting.
```
Comment (hmistry) [inline text/0002-bundle-add-and-remove.md]: I'd suggest changing `aborting` to `skipping`. Aborting sounds very serious and I wouldn't want it to abort in multi gem additions. I prefer it skips and adds remaining gems

File: text/0002-bundle-add-and-remove.md [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+
+### 7.2. removing gem when it's not present
+
+```ruby
+# Gemfile
+source "https://rubygems.org"
+```
+
+```
+bundle rm minitest sinatra
```
Comment (hmistry) [inline text/0002-bundle-add-and-remove.md]: `bundle remove` if all agrees

File: text/0002-bundle-add-and-remove.md [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+```
+
+Prints out and exits with error status code:
+
+```
+Gem "minitest" couldn't be found in the Gemfile. Aborting.
```
Comment (hmistry) [inline text/0002-bundle-add-and-remove.md]: Same comment as before `skipping` over `aborting`.

File: text/0002-bundle-add-and-remove.md [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+which are more natural for users.
+
+It appends gems to the end of the group block if the group already exists. It also supports adding to multiple groups.
+
+## 2. adding multiple gems with options
+
+#### Proposed command
+
+```shell
+bundle add sinatra:2.0 minitest:~>5.11
```
Comment (booch) [inline text/0002-bundle-add-and-remove.md]: You're going to need quotes around the 2nd gem there, or the shell will interpret the `>` as output redirection. I'd recommend putting single quotes around all the gem names/versions in shell commands you show in the docs.

**Comment by shime** [comment]
Thanks for your input, everyone! I've made the updates to the RFC document. Regarding @segiddins's comment - I don't think we can do much when it comes to arbitrary Ruby code inside a Gemfile. 

These commands would operate on the text in the Gemfile, they would not execute any Ruby code inside it. So for removing, it would report failure to do so if a gem is not found in a Gemfile text. Same goes for adding, it would check if a gem is present in a Gemfile text and if not it would try to add it to according group.

**Comment by indirect** [comment]
Accepting this PR, since implementation has already begun as part of GSoC 2018. 😄

**Pull request merged**